ipn/ipnlocal/local.go

```diff
@@ -267,9 +269,56 @@ func NewLocalBackend(logf logger.Logf, logid string, store ipn.StateStore, diale
 		b.logf("[unexpected] failed to wire up peer API port for engine %T", e)
 	}
 
+	if v, err := store.ReadState(ipn.DebugMagicsockUntilStateKey); err == nil {
+		if ut, err := strconv.ParseInt(string(v), 10, 64); err == nil {
+			b.SetComponentDebugLogging("magicsock", time.Unix(ut, 0))
+		}
+	}
+
 	return b, nil
 }
 
+type componentLogState struct {
+	until time.Time
+	timer *time.Timer // if non-nil, the AfterFunc to disable it
+}
+
+// SetComponentDebugLogging sets component's debug logging enabled until the until time.
+// If until is in the past, the component's debug logging is disabled.
+//
+// The following components are recognized:
+//
+//   - magicsock
+func (b *LocalBackend) SetComponentDebugLogging(component string, until time.Time) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+	switch component {
+	default:
+		return fmt.Errorf("unknown component %q", component)
+	case "magicsock":
+		mc, err := b.magicConn()
+		if err != nil {
+			return err
+		}
+		newSt := componentLogState{until: until}
+		on := now.Before(until)
+		mc.SetDebugLoggingEnabled(on)
+
+		if oldSt, ok := b.componentLogUntil[component]; ok && oldSt.timer != nil {
+			oldSt.timer.Stop()
```

Comment:
Are we okay with the (presumably-unlikely) case where we fail to stop the timer because it's in-flight?

Reply:
I'm thinking of the following sequence:

**T -10:** `SetComponentDebugLogging(component, T=0)`
**T 0:** timer.AfterFunc starts
**T 0.1:** `SetComponentDebugLogging(component, T=5)`, which sets debugging on, and calls `timer.Stop()` which returns false
**T 0.2:** timer.AfterFunc runs, sets debugging off

Now it looks like we enabled debug logging, but it's actually off.

Reply:
(this is also super unlikely, so probably worth ignoring)

Reply:
Fixed, I think. PTAL in any case.

Reply:
Yep, I think that fixes it :+1: Thanks!